internal/services/server.go

```diff
@@ -62,4 +69,35 @@ func RegisterGrpcServices(
 	healthpb.RegisterHealthServer(srv, healthSrv)
 
 	reflection.Register(grpcutil.NewAuthlessReflectionInterceptor(srv))
+
+	go checkDispatcherServicesReady(dispatch, healthSrv, v1.PermissionsService_ServiceDesc.ServiceName, OverallServerHealthCheckKey)
+}
+
+// checkDispatcherServicesReady waits for the dispatcher to become ready before setting the health check status for the given services
+func checkDispatcherServicesReady(dispatch dispatch.Dispatcher, healthSrv *grpcutil.AuthlessHealthServer, services ...string) {
+	backoffInterval := backoff.NewExponentialBackOff()
+	// Run immediately for the initial check
+	ticker := time.After(0)
+
+	for {
```

Comment:
there's also https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange which you could use to detect immediately when there's a change in connection state?

Reply:
+1 to both points